flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobResourceRequirements.java

```diff
@@ -0,0 +1,222 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.jobgraph;
+
+import org.apache.flink.util.InstantiationUtil;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Information about the parallelism of job vertices. */
+public class JobResourceRequirements implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * A key for an internal config option (intentionally prefixed with $internal to make this
+     * explicit), that we'll serialize the {@link JobResourceRequirements} into, when writing it to
+     * {@link JobGraph}.
+     */
+    private static final String JOB_RESOURCE_REQUIREMENTS_KEY =
+            "$internal.job-resource-requirements";
+
+    private static final JobResourceRequirements EMPTY =
+            new JobResourceRequirements(Collections.emptyMap());
+
+    /**
+     * Write {@link JobResourceRequirements resource requirements} into the configuration of a given
+     * {@link JobGraph}.
+     *
+     * @param jobGraph job graph to write requirements to
+     * @param jobResourceRequirements resource requirements to write
+     * @throws IOException in case we're not able to serialize requirements into the configuration
+     */
+    public static void writeToJobGraph(
+            JobGraph jobGraph, JobResourceRequirements jobResourceRequirements) throws IOException {
+        InstantiationUtil.writeObjectToConfig(
+                jobResourceRequirements,
+                jobGraph.getJobConfiguration(),
+                JOB_RESOURCE_REQUIREMENTS_KEY);
+    }
+
+    /**
+     * Read {@link JobResourceRequirements resource requirements} from the configuration of a given
+     * {@link JobGraph}.
+     *
+     * @param jobGraph job graph to read requirements from
+     * @throws IOException in case we're not able to deserialize requirements from the configuration
+     * @throws ClassNotFoundException in case some deserialized classes are missing on the classpath
+     */
+    public static Optional<JobResourceRequirements> readFromJobGraph(JobGraph jobGraph)
+            throws IOException, ClassNotFoundException {
```

Comment:
nit: `ClassNotFoundException` doesn't seem appropriate at this level, would it make sense to wrap it into `IOException` (or some runtime)?

Reply:
> I guess the only way to throw a ClassNotFoundException here is for JobResourceRequirements (or inner ones) class to be missing ?

Or if the JRR was written by a different Flink version (this method will be used for reading from a JobGrapStore).

Reply:
@rkhachatryan Thinking more about the reason I've described above, I think it's legit (it's; we can run in this case during failover (if we'd fail over to a different JM version).

All in all, it's just a shorthand for calling `InstantiationUtil.readObjectFromConfig`, and all the usages already throw the checked exceptions anyway, so it doesn't add any additional overhead.

Reply:
I still think that `ClassNotFoundException` is not the right exception to throw in case of versions incompatibility, but probably it's subjective. And as you said, it doesn't require any special handling by clients :+1:

Reply:
I've added a wrapper with an explainer why this might happen.